Msft document intelligence ocr

gordl opened the pull request:
## Change Description

Initial support for document intelligence

## Issue reference

This PR fixes issue #1183 

## Checklist

- [x] I have reviewed the [contribution guidelines](https://github.com/microsoft/presidio/blob/main/CONTRIBUTING.md)
- [x] I have signed the CLA (if required)
- [x] My code includes unit tests
- [x] All unit tests and lint checks pass locally
- [x] My PR contains documentation updates / additions if required

gordl commented:
> @gordl please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement
> ### Contribution License Agreement
> This Contribution License Agreement (**“Agreement”**) is agreed to by the party signing below (**“You”**), and conveys certain license rights to Microsoft Corporation and its affiliates (“Microsoft”) for Your contributions to Microsoft open source projects. This Agreement is effective as of the latest signature date below.
> 
> 1. **Definitions**.
>    **“Code”** means the computer software code, whether in human-readable or machine-executable form,
>    that is delivered by You to Microsoft under this Agreement.
>    **“Project”** means any of the projects owned or managed by Microsoft and offered under a license
>    approved by the Open Source Initiative ([www.opensource.org](http://www.opensource.org)).
>    **“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
>    Project, including but not limited to communication on electronic mailing lists, source code control
>    systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
>    discussing and improving that Project, but excluding communication that is conspicuously marked or
>    otherwise designated in writing by You as “Not a Submission.”
>    **“Submission”** means the Code and any other copyrightable material Submitted by You, including any
>    associated comments and documentation.
> 2. **Your Submission**. You must agree to the terms of this Agreement before making a Submission to any
>    Project. This Agreement covers any and all Submissions that You, now or in the future (except as
>    described in Section 4 below), Submit to any Project.
> 3. **Originality of Work**. You represent that each of Your Submissions is entirely Your original work.
>    Should You wish to Submit materials that are not Your original work, You may Submit them separately
>    to the Project if You (a) retain all copyright and license information that was in the materials as You
>    received them, (b) in the description accompanying Your Submission, include the phrase “Submission
>    containing materials of a third party:” followed by the names of the third party and any licenses or other
>    restrictions of which You are aware, and (c) follow any other instructions in the Project’s written
>    guidelines concerning Submissions.
> 4. **Your Employer**. References to “employer” in this Agreement include Your employer or anyone else
>    for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
>    Submission is made in the course of Your work for an employer or Your employer has intellectual
>    property rights in Your Submission by contract or applicable law, You must secure permission from Your
>    employer to make the Submission before signing this Agreement. In that case, the term “You” in this
>    Agreement will refer to You and the employer collectively. If You change employers in the future and
>    desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
>    and secure permission from the new employer before Submitting those Submissions.
> 5. **Licenses**.
> 
> * **Copyright License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license in the
>   Submission to reproduce, prepare derivative works of, publicly display, publicly perform, and distribute
>   the Submission and such derivative works, and to sublicense any or all of the foregoing rights to third
>   parties.
> * **Patent License**. You grant Microsoft, and those who receive the Submission directly or
>   indirectly from Microsoft, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license under
>   Your patent claims that are necessarily infringed by the Submission or the combination of the
>   Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
>   import or otherwise dispose of the Submission alone or with the Project.
> * **Other Rights Reserved**. Each party reserves all rights not expressly granted in this Agreement.
>   No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
>   granted by implication, exhaustion, estoppel or otherwise.
> 
> 6. **Representations and Warranties**. You represent that You are legally entitled to grant the above
>    licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
>    have disclosed under Section 3). You represent that You have secured permission from Your employer to
>    make the Submission in cases where Your Submission is made in the course of Your work for Your
>    employer or Your employer has intellectual property rights in Your Submission by contract or applicable
>    law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
>    have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
>    You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
>    REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
>    EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
>    PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
>    NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> 7. **Notice to Microsoft**. You agree to notify Microsoft in writing of any facts or circumstances of which
>    You later become aware that would make Your representations in this Agreement inaccurate in any
>    respect.
> 8. **Information about Submissions**. You agree that contributions to Projects and information about
>    contributions may be maintained indefinitely and disclosed publicly, including Your name and other
>    information that You submit with Your Submission.
> 9. **Governing Law/Jurisdiction**. This Agreement is governed by the laws of the State of Washington, and
>    the parties consent to exclusive jurisdiction and venue in the federal courts sitting in King County,
>    Washington, unless no federal subject matter jurisdiction exists, in which case the parties consent to
>    exclusive jurisdiction and venue in the Superior Court of King County, Washington. The parties waive all
>    defenses of lack of personal jurisdiction and forum non-conveniens.
> 10. **Entire Agreement/Assignment**. This Agreement is the entire agreement between the parties, and
>     supersedes any and all prior agreements, understandings or communications, written or oral, between
>     the parties relating to the subject matter hereof. This Agreement may be assigned by Microsoft.

@microsoft-github-policy-service agree company="Microsoft"

omri374 commented:
Thanks @gordl! this is a great addition.

omri374 commented on presidio-image-redactor/tests/test_document_intelligence_ocr.py:
```diff
@@ -0,0 +1,78 @@
+
+
+#Mark this test as optionally failing
+@pytest.mark.xfail(reason="This test is expected to fail unless the environment variables are set")
```
Instead of xfail, can we consider using `pytest.skip` or a condition inside the test checking if environment variables are available?

gordl replied:
Good suggestion.  I made some changes so that constructor failure results in a test skip.

omri374 commented on presidio-image-redactor/tests/integration/test_dicom_image_redactor_engine_integration.py:
```diff
@@ -33,33 +48,34 @@ def mock_engine():
-def test_redact_image_correctly(
-    mock_engine: DicomImageRedactorEngine, dcm_filepath: Path
-):
+@pytest.mark.parametrize("engine_builder", all_engines_required())
```
Cool!

gordl commented on presidio-image-redactor/presidio_image_redactor/document_intelligence_ocr.py:
```diff
@@ -0,0 +1,115 @@
+        return poller.result()
+
+    def perform_ocr(self, image:object, **kwargs) -> dict:
+        """Perform OCR on the image"""
```
You're right, there is much more that could be done.  What I've done is added the ability to pass arbitrary parameters to the client.  To your example, not all engines support key/value pairs, and many of the supported engines support different kinds of return values.  What is consistent, is that they all support document->page->word so I've left that method of interpretation as the standard.  We (or a user) could certainly expand upon this in the future.

omri374 replied:
Is there a way to specify that there are additional ways to implement Document Intelligence? I'm concerned that users would say "DI doesn't work for me" without realizing they can customize it to their needs. Even a comment in code (and/or in the docs) describing the implementation in high level could be useful. We can add a note saying that the current implementation treats words as bounding boxes, but there are additional ways one can call DI for PHI/PII detection purposes.

gordl replied:
Added a high level overview in the documentation.

gordl commented on presidio-image-redactor/presidio_image_redactor/document_intelligence_ocr.py:
```diff
@@ -0,0 +1,115 @@
+        "prebuilt-businessCard"
+    ]
+
+    def __init__(self, endpoint = None, key = None, model_id = "prebuilt-document"):
```
Fixed in most recent iteration.

gordl commented on presidio-image-redactor/presidio_image_redactor/document_intelligence_ocr.py:
```diff
@@ -0,0 +1,115 @@
+        self.model_id = model_id
+
+    @staticmethod
+    def _polygon_to_bbox(polygon):
```
See most recent commit

gordl commented on presidio-image-redactor/Pipfile:
```diff
@@ -14,6 +14,7 @@ python-gdcm = ">=3.0.22,<4.0.0"
 matplotlib = ">=3.6.2,<4.0.0"
 opencv-python = ">=4.8.0"
 typing-extensions = "*"
+azure-ai-formrecognizer = "*"
```
done.

gordl commented:
This is ready for a second look.  

Major changes since the last review including marking as 'skip' any tests that require a key to the DI OCR engine instead of xfail.  Providing an incorrect key results in a test failure, all tests pass if you provide a key.  

Second change I made was to also test the image redactor engine in addition to dicom image redactor, and in order to accommodate that I made some changes to the how tests pass based on image similarity because of different ocr behaviours.

omri374 commented:
/azp run

omri374 reviewed (dismissed):
Looks great, thanks!

niwilso commented on presidio-image-redactor/tests/integration/test_dicom_image_redactor_engine_integration.py:
```diff
@@ -33,33 +57,33 @@ def mock_engine():
 
     Args:
-        mock_engine (DicomImageRedactorEngine): Mock instance.
+        engine (DicomImageRedactorEngine): Mock instance.
```
**nit**: Typo "engine" -> "engine_builder"

niwilso replied:
I recommend doing a ctrl + F since this occurs in multiple docstrings in this file.

niwilso reviewed (changes requested):
Great addition to Presidio! Just a few small changes that should be done prior to merge.

gordl commented:
@niwilso Thank you for your review!  I've gone ahead and resolved those comments.

omri374 commented:
/azp run
